### ISSUE TITLE
How to make a shared reference to a display?

### ISSUE THREAD

**Issue by deltronix** [Finomnis/st7565#17]
As the title says, how would I go about making a shared reference to a display. I cannot seem to make a Trait object of DrawableTarget or a shared reference to the ST7565 struct.

**Comment by Finomnis** [comment]
Hey! Thanks for your interest in this crate.

I'm having a hard time following what you mean with a "shared reference" in this case. All of the drawing functions require `mut`, so they require exclusive access.

Which system are you using? A direct `main` based program? An rtic based program?

It will probably require some kind of locking mechanism based on critical sections, and putting the driver somewhere in a static memory. But that strongly depends on your program, so I might have to see a minimal example of what you are trying to achieve.

**Comment by deltronix** [comment]
Hey @Finomnis , thanks for the response, I should have been more clear on my intent and the context of my question.

I'm writing an RTIC (v2) based program and would like to use the display driver (and the GraphicsPageBuffer) in a task (preferrably in the shared context). However I am running into lifetime issues and am not sure how to statically borrow the driver.

Should I use something like heapless::pool to store the driver and page buffer?

Here is my attempt so far (I have made GraphicsMode public in the driver to achive this).

```rust
#![deny(unsafe_code)]
#![no_main]
#![no_std]
#![feature(type_alias_impl_trait)]

use control_module as _; // global logger + panicking-behavior + memory layout
use stm32f4xx_hal::prelude::*;
use rtic::app;
use embedded_hal::spi::{MODE_3};
use display_interface_spi::SPIInterface;
use embedded_graphics::{
    geometry::Size,
    mono_font::{ascii::FONT_8X13, MonoTextStyle},
    pixelcolor::BinaryColor,
    prelude::*,
    primitives::{Circle, PrimitiveStyle, Rectangle},
    text::Text,
};
use st7565::{displays::DOGL128_6_EXT12V, GraphicsPageBuffer, ST7565, GraphicsMode};
use stm32f4xx_hal::{
    spi::Spi,
    spi::Spi2,
    gpio::{PB11, PB12},
};
use rtic_monotonics::systick::*;
use heapless::pool::Box;

#[app(
    device = stm32f4xx_hal::pac,
    peripherals = true,
    dispatchers = [TIM3]
)]
mod app {

    use stm32f4xx_hal::gpio::{Pin, Output};

    use super::*;
    //use stm32f4xx_hal::gpio::gpiob::{PB10, PB11, PB12, PB13, PB14, PB15};
    // Shared resources go here
    //
    //
    type DisplayType = ST7565<
        SPIInterface<Spi2,Pin<'B',11, Output>,Pin<'B',12, Output>>,
        DOGL128_6_EXT12V,
        GraphicsMode<'static, 128, 8>,
        128,64,8>;
    #[shared]
    struct Shared {
    }

    // Local resources go here
    #[local]
    struct Local {
        disp: &'static DisplayType,
        // TODO: Add resources
    }

    #[init]
    fn init(cx: init::Context) -> (Shared, Local) {
        defmt::info!("init");
        defmt::info!("RCC constrain");

        let mut rcc = cx.device.RCC.constrain();

        let systick_mono_token = rtic_monotonics::create_systick_token!();
        Systick::start(cx.core.SYST, 48_000_000, systick_mono_token); 

        let ccdr = rcc.cfgr.sysclk(48.MHz()).freeze();
        // defmt::info!("RCC constrain")
        //let syscfg = cx.device.SYSCFG.constrain();

        defmt::info!("gpio setup");
        let (mut spi2_disp_rst, mut spi2_disp_a0, mut spi2_disp_cs, spi2_sck, spi2_miso, spi2_mosi, mut disp_backlight) = {
            let gpiob = cx.device.GPIOB.split();
            (gpiob.pb10.into_push_pull_output(), 
             gpiob.pb11.into_push_pull_output(), 
             gpiob.pb12.into_push_pull_output(), 
             gpiob.pb13.into_alternate(), 
             gpiob.pb14.into_alternate(), 
             gpiob.pb15.into_alternate(),
             gpiob.pb1.into_push_pull_output(),
             )
        };
        disp_backlight.set_high();
        let mut spi2_flash_cs = {
            let gpioc = cx.device.GPIOC.split();
            gpioc.pc6.into_push_pull_output()
        };

        spi2_flash_cs.set_high();
        spi2_disp_rst.set_high();
        spi2_disp_a0.set_high();
        spi2_disp_cs.set_high();


        defmt::info!("display init");
        let disp_spi = 
            SPIInterface::new(
            Spi::new(
                cx.device.SPI2,
                (spi2_sck, spi2_miso, spi2_mosi),
                MODE_3,
                1.MHz(),
                &ccdr
            ),
            spi2_disp_a0,
            spi2_disp_cs,
            );

        let mut timer = cx.device.TIM2.delay_us(&ccdr);

        let mut page_buffer = GraphicsPageBuffer::new();
        let mut disp = ST7565::new(disp_spi, DOGL128_6_EXT12V).into_graphics_mode(&mut page_buffer);
        disp.reset(&mut spi2_disp_rst, &mut timer).unwrap();
        disp.flush().unwrap();
        disp.set_display_on(true).unwrap();

        Circle::new(Point::new(6, 6), 20)
            .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 2))
            .draw(&mut disp)
            .unwrap();
        Rectangle::new(Point::new(106, 6), Size::new(20, 20))
            .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 2))
            .draw(&mut disp)
            .unwrap();
        let font = MonoTextStyle::new(&FONT_8X13, BinaryColor::On);
        Text::new("Hello,\nRust!", Point::new(43, 13), font)
            .draw(&mut disp)
            .unwrap();

        // Send content to display
        task1::spawn().ok();

        (
            Shared {
            },
            Local {
                // Initialization of local resources go here
                disp: &disp,
            },
        )
    }

    // Optional idle, can be removed if not needed.
    #[idle]
    fn idle(_: idle::Context) -> ! {
        defmt::info!("idle");

        loop {
            continue;
        }
    }

    // TODO: Add tasks
    #[task(priority = 1)]
    async fn task1(_cx: task1::Context) {
        defmt::info!("Hello from task1!");
    }
}

```

**Comment by Finomnis** [comment]
Do you have a repository where I can just hit `cargo build` to reproduce your error? I obviously don't have the hardware, but it would be very tedious to collect all the `Cargo.toml` and other files like `rust-toolchain` or similar that are required to cross-compile this. If you give me something like that I am confident I can quickly show you a way.

You are almost there - you are only missing two things. For one, if you want to share it with multiple contexts, you can't have it in `Local` - you need to put it in `Shared`, so it gets auto-wrapped in a mutex.

The second one is the question of how to actually create a `&'static mut` reference - and for that, RTIC has its own dedicated way. It uses `Option<...>` arguments as **input** parameters of `init()` to achieve that. This is teased a little bit in the [rtic book](https://rtic.rs/2/book/en/by-example/tips/static_lifetimes.html), but I think it's easier if I can show you, once you provide me with a minimal reproducible repository that I can modify.

The third thing you will stumble across is similar to the second one - the pagebuffer will also want to be a `&'static mut` reference. This will be solved identically - with an `Option` input parameter.

The trick is that you do `#[init(local = [x: Option<T> = None])]`, and then inside the `init` function you create a `T`. Then, you can convert the `T` to a `&'static mut T` via the [`Option::insert`](https://doc.rust-lang.org/std/option/enum.Option.html#method.insert) method.

**Comment by deltronix** [comment]
Ahh I think I can see where you are going! I have uploaded my code [here](https://github.com/deltronix/control_module/) for you to check out.

**Comment by Finomnis** [comment]
Also, good catch that the modes should be public. Otherwise it's impossible to create a shared object in RTIC. Feel free to pull request your changes once your display works.

**Comment by Finomnis** [comment]
@deltronix This compiles:

```rust
#![deny(unsafe_code)]
#![no_main]
#![no_std]
#![feature(type_alias_impl_trait)]

use control_module as _; // global logger + panicking-behavior + memory layout
use display_interface_spi::SPIInterface;
use embedded_graphics::{
    geometry::Size,
    mono_font::{ascii::FONT_8X13, MonoTextStyle},
    pixelcolor::BinaryColor,
    prelude::*,
    primitives::{Circle, PrimitiveStyle, Rectangle},
    text::Text,
};
use embedded_hal::spi::MODE_3;
use rtic::app;
use rtic_monotonics::systick::*;
use st7565::{displays::DOGL128_6_EXT12V, GraphicsMode, GraphicsPageBuffer, ST7565};
use stm32f4xx_hal::prelude::*;
use stm32f4xx_hal::{spi::Spi, spi::Spi2};

#[app(
    device = stm32f4xx_hal::pac,
    peripherals = true,
    dispatchers = [TIM3]
)]
mod app {

    use stm32f4xx_hal::gpio::{Output, Pin};

    use super::*;
    //use stm32f4xx_hal::gpio::gpiob::{PB10, PB11, PB12, PB13, PB14, PB15};
    // Shared resources go here
    //
    //
    type DisplayType = ST7565<
        SPIInterface<Spi2, Pin<'B', 11, Output>, Pin<'B', 12, Output>>,
        DOGL128_6_EXT12V,
        GraphicsMode<'static, 128, 8>,
        128,
        64,
        8,
    >;
    #[shared]
    struct Shared {
        disp: &'static mut DisplayType,
    }

    // Local resources go here
    #[local]
    struct Local {
        // TODO: Add resources
    }

    #[init(local = [
        disp: Option<DisplayType> = None,
        page_buffer: Option<GraphicsPageBuffer<128,8>> = None
    ])]
    fn init(cx: init::Context) -> (Shared, Local) {
        defmt::info!("init");
        defmt::info!("RCC constrain");

        let mut rcc = cx.device.RCC.constrain();

        let systick_mono_token = rtic_monotonics::create_systick_token!();
        Systick::start(cx.core.SYST, 48_000_000, systick_mono_token);

        let ccdr = rcc.cfgr.sysclk(48.MHz()).freeze();
        // defmt::info!("RCC constrain")
        //let syscfg = cx.device.SYSCFG.constrain();

        defmt::info!("gpio setup");
        let (
            mut spi2_disp_rst,
            mut spi2_disp_a0,
            mut spi2_disp_cs,
            spi2_sck,
            spi2_miso,
            spi2_mosi,
            mut disp_backlight,
        ) = {
            let gpiob = cx.device.GPIOB.split();
            (
                gpiob.pb10.into_push_pull_output(),
                gpiob.pb11.into_push_pull_output(),
                gpiob.pb12.into_push_pull_output(),
                gpiob.pb13.into_alternate(),
                gpiob.pb14.into_alternate(),
                gpiob.pb15.into_alternate(),
                gpiob.pb1.into_push_pull_output(),
            )
        };
        disp_backlight.set_high();
        let mut spi2_flash_cs = {
            let gpioc = cx.device.GPIOC.split();
            gpioc.pc6.into_push_pull_output()
        };

        spi2_flash_cs.set_high();
        spi2_disp_rst.set_high();
        spi2_disp_a0.set_high();
        spi2_disp_cs.set_high();

        defmt::info!("display init");
        let disp_spi = SPIInterface::new(
            Spi::new(
                cx.device.SPI2,
                (spi2_sck, spi2_miso, spi2_mosi),
                MODE_3,
                1.MHz(),
                &ccdr,
            ),
            spi2_disp_a0,
            spi2_disp_cs,
        );

        let mut timer = cx.device.TIM2.delay_us(&ccdr);

        let page_buffer = cx.local.page_buffer.insert(GraphicsPageBuffer::new());
        let mut disp = ST7565::new(disp_spi, DOGL128_6_EXT12V).into_graphics_mode(page_buffer);
        disp.reset(&mut spi2_disp_rst, &mut timer).unwrap();
        disp.flush().unwrap();
        disp.set_display_on(true).unwrap();
        let disp = cx.local.disp.insert(disp);

        Circle::new(Point::new(6, 6), 20)
            .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 2))
            .draw(disp)
            .unwrap();
        Rectangle::new(Point::new(106, 6), Size::new(20, 20))
            .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 2))
            .draw(disp)
            .unwrap();
        let font = MonoTextStyle::new(&FONT_8X13, BinaryColor::On);
        Text::new("Hello,\nRust!", Point::new(43, 13), font)
            .draw(disp)
            .unwrap();

        // Send content to display
        task1::spawn().ok();

        (Shared { disp }, Local {})
    }

    // Optional idle, can be removed if not needed.
    #[idle]
    fn idle(_: idle::Context) -> ! {
        defmt::info!("idle");

        loop {
            continue;
        }
    }

    // TODO: Add tasks
    #[task(priority = 1, shared=[disp])]
    async fn task1(mut cx: task1::Context) {
        defmt::info!("Hello from task1!");

        cx.shared.disp.lock(|disp| {
            disp.flush().unwrap();
        });
    }
}
```

**Comment by deltronix** [comment]
> @deltronix This compiles:

It works on the hardware as well, thanks!



> Also, good catch that the modes should be public. Otherwise it's impossible to create a shared object in RTIC. Feel free to pull request your changes once your display works.

Yeah will do so now. When I've fleshed out my RTIC code a bit more I will add it to the examples as well.

**Comment by Finomnis** [comment]
@deltronix Thanks for your contribution :)

I also wrote a non-monospace font renderer for embedded-graphics, which you might find useful for your project: https://github.com/Finomnis/u8g2-fonts

Greetings :)

**Comment by Finomnis** [comment]
Published as v0.3.0 :)

**Comment by Finomnis** [comment]
I'll close this, if you have any questions feel free to ask anyway.
